### PR TITLE
Initial cut for Shark-64

### PR DESCRIPTION
**Pull request by harsha2010**
https://spark-project.atlassian.net/browse/SHARK-64
This checkin fixes Shark-64.
When cached tables are created, meta data is added to the hive meta store for each table, that includes the query that resulted in the cached table creation.
Upon server restarts, we run through the queries to re populate the cached RDDs before resuming work.
Also added a few JDBC tests for Shark Server

### CODE REVIEW

File: src/main/scala/shark/SharkServer.scala [hunk 1]
```diff
@@ -43,24 +58,29 @@ object SharkServer {
...
+      while(!server.isServing()) {}
+      val sshandler = new SharkServerHandler
+      SharkCTAS.loadAsRdds(sshandler.execute(_))
+    }
     LOG.info("Starting shark server on port " + port)
     server.serve()
   }
-
+  
+  def stop = if(server !=null) server.stop
```
Comment (rxin) [inline src/main/scala/shark/SharkServer.scala]: nitpick: a space after if.

File: src/main/scala/shark/SharkServer.scala [hunk 2]
```diff
@@ -43,24 +58,29 @@ object SharkServer {
...
-          SharkEnv.stop()
+          SharkServer.stop
         }
       }
     )
 
+    spawn {
+      while(!server.isServing()) {}
+      val sshandler = new SharkServerHandler
+      SharkCTAS.loadAsRdds(sshandler.execute(_))
```
Comment (rxin) [inline src/main/scala/shark/SharkServer.scala]: What's the behavior if users are querying these cached tables when the server is first started? Should we block the server when it is creating cached tables?

**Comment by harsha2010** [comment]
Thanks Reynold, will fix the space and recommit.
Yes, you are right, we will need to block the server when it is creating cached tables. I will add that in and check in

**Comment by rxin** [comment]
BTW - it might be better to have a flag in the server/cli start script to
disable the loading, in case the loading itself can deterministically go
wrong.

## 

Reynold Xin
AMPLab, UC Berkeley
http://www.cs.berkeley.edu/~rxin/

**Comment by harsha2010** [comment]
good point, will include it. I also need to harden the code to deal with partial failures : if reloading one of several cached tables fail, we should log it and move on, i would think..but the patch exits eagerly.

File: src/main/scala/shark/SharkServer.scala [hunk 3]
```diff
@@ -27,46 +38,83 @@ object SharkServer {
...
       }
     )
-
+    execLoadRdds(loadRdds, latch)
     LOG.info("Starting shark server on port " + port)
     server.serve()
   }
-
+  
+  def stop = if (server != null) {
```
Comment (rxin) [inline src/main/scala/shark/SharkServer.scala]: Also need to call SharkEnv.sc.stop() here.

File: src/main/scala/shark/SharkServer.scala [hunk 4]
```diff
@@ -27,46 +38,83 @@ object SharkServer {
...
+      try {
+        val sshandler = new SharkServerHandler
+        SharkCTAS.loadAsRdds(sshandler.execute(_))
+        LOG.info("Executed load "+ SharkCTAS.getMeta)
+      } catch {
+        case (e: Exception) => LOG.warn("Unable to load RDDs upon startup", e)
+      } finally {
+        latch.countDown
+      }
+  }
```
Comment (rxin) [inline src/main/scala/shark/SharkServer.scala]: The brackets here are correct, but look a little bit confusing. Can we add more parenthesis to make it more clear? 

E.g.

```
if (!loadFlag) { 
  latch.countDown
} else spawn {
  ...
}
```

File: src/main/scala/shark/memstore/CacheManager.scala [hunk 5]
```diff
@@ -23,4 +23,10 @@ class CacheManager {
     keyToRdd.keys.map(_.key).collect { case k: String => k } toSeq
   }
 
+  def drop(): Seq[String] = {
```
Comment (rxin) [inline src/main/scala/shark/memstore/CacheManager.scala:26]: Is this only used for testing? If yes, can we add a comment saying that?

**Comment by rxin** [comment]
Closing this one because I am merging it manually.

**Pull request closed without merging**